exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java

```diff
@@ -146,19 +146,25 @@ public boolean isPositive() {
 
     @Override
     protected @Nullable IntSupplier createComparisonWith(final NumericValue other) {
-        if (other instanceof IntegerValue) {
-            return () -> BigDecimal.valueOf(value).compareTo(new BigDecimal(((IntegerValue) other).value));
-        }
-        if (other instanceof DecimalValue) {
-            return () -> BigDecimal.valueOf(value).compareTo(((DecimalValue) other).value);
-        }
-        if (other instanceof DoubleValue) {
-            return () -> Double.compare(value, ((DoubleValue) other).value);
-        }
-        if (other instanceof FloatValue) {
-            return () -> Double.compare(value, ((FloatValue) other).value);
-        }
-        return null;
+        final IntSupplier comparison;
+        if (isNaN()) {
+            comparison = () -> Constants.INFERIOR;
+        } else if (other.isNaN()) {
+            comparison = () -> Constants.SUPERIOR;
+        } else if (isInfinite() && other.isInfinite() && isPositive() == other.isPositive()) {
+            comparison = () -> Constants.EQUAL;
+        } else if (other instanceof IntegerValue) {
+            comparison = () -> BigDecimal.valueOf(value).compareTo(new BigDecimal(((IntegerValue) other).value));
+        } else if (other instanceof DecimalValue) {
+            comparison = () -> BigDecimal.valueOf(value).compareTo(((DecimalValue) other).value);
+        } else if (other instanceof DoubleValue) {
+            comparison = () -> Double.compare(value, ((DoubleValue) other).value);
+        } else if (other instanceof FloatValue) {
+            comparison = () -> Double.compare(value, ((FloatValue) other).value);
```

Comment:
The enhanced `instanceof` with direct cast should be used here

Reply:
Done.